components/tel/Core/Src/xbee.c

```diff
@@ -0,0 +1,87 @@
+/**
+ *  @file xbee.c
+ *  @brief Defines functions related to xBee module, including setting up xBee packet overhead, and handling return frames
+ *
+ *  @date 2024/06/16
+ *  @author Evan Owens
+ */
+
+#include "xbee.h"
+#include <math.h>
+
+/**
+ * @brief Adds up all the bytes after length and before checksum in an api packet
+ *
+ * @param api packet: api packet array
+ * @return uint16_t total_bytes in api packet before checksum and after length fields
+*/
+
+uint16_t XBEE_sum_bytes (uint8_t api_packet[], uint16_t packet_length)
+{
+  uint16_t total_bytes = 0;
+  for (int i = 3; i < packet_length; i++){
+		 total_bytes += api_packet[i];
+  }
+  return total_bytes;
+}
+
+
+/**
+ * @brief Sets up overhead for an API Packet
+ *
+ * @param api packet: api packet array
+ * @return api_packet with setup api overhead
+*/
+void XBEE_api_overhead_setup(uint8_t api_packet[])
+{
+  uint8_t bit_address_64[] = {0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00};
+
+  api_packet[START_DELIMITER_POSITION] = START_DELIMITER;
+  api_packet[FRAME_TYPE_POSITION] = FRAME_TYPE;
+  api_packet[FRAME_ID_POSITION] = FRAME_ID;
+  api_packet[BIT_ADDRESS_16_LOW_POSITION] = BIT_ADDRESS_16_LOW;
+  api_packet[BIT_ADDRESS_16_HIGH_POSITION] = BIT_ADDRESS_16_HIGH;
+  api_packet[BROADCAST_RADIUS_POSITION] = BROADCAST_RADIUS;
+  api_packet[OPTIONS_POSITION] = OPTIONS;
+
+  for (int i =0; i < 8; i++){
+	  api_packet[i] = bit_address_64[i];
+  }
+
+}
+
+
+/**
+ * @brief Calculates and adds the MSB and LSB for an API packet
+ *
+ * @param api packet: api packet array
+ * @return api_packet with setup length fields
+*/
+void XBEE_calculate_length(uint8_t api_packet[], uint16_t packet_length)
+{
+ uint16_t total_bytes = packet_length - CHECKSUM_LENGTH_SIZE;
```

Comment:
Why do you do packet_length - CHECKSUM_LENGTH_SIZE? The packet length does not seem to include the checksum length in your code so it would be the equivalent of taking of CHECKSUM_LENGTH_SIZE bytes from the **original** packet.

Reply:
Packet length should include the checksum (api_packet_size = API_OVERHEAD_SIZE + api_buffer_position) with API_OVERHEAD_SIZE including the checksum. What this is just doing is removing all the bytes from the start delimiter, length fields, and checksum so that length can be properly calculated.

Reply:
When you say API_OVERHEAD_SIZE including the checksum, where do you set the checksum in the api_buffer? Is this something that is not implemented yet or is it written to the API buffer elsewhere. The same applies for the length field. I ask this because i dont think I see it in the overhead_setup function here:
![image](https://github.com/UBC-Solar/firmware_v3/assets/117491745/8dfc5671-a88c-48c9-b278-f7e043c90fb2)

Reply:
API_OVERHEAD_SIZE is a constant and its just the number of bytes of api overhead in every frame. Checksum is set by the XBEE_calculate_checksum(api_packet, api_packet_size); in the  RADIO_tx_API_Packager function

Reply:
length is also set in the radio_tx_api_packager function by a function i wrote in the xbee.c file

Reply:
same here, any relevant details should be mentioned in the function bride or as comments for readability.